### PR TITLE
Version check Yuhang Liao

### PR DESCRIPTION
**Pull request by ComarPers922**
UPDATE FROM YUHANG LIAO

1. Version Check feature is added. Type "info" to get current version information.

2. Modify some error messages.

The previous pull request has been closed and invalidated.

### CODE REVIEW

**Comment by lepidopterane-atsmith** [comment]
This Code Review is going to be a lot like the last one. Of course reach out to me on Slack! **I only have a few bolded questions which I expect you to answer on our channel.**

**Comment by lepidopterane-atsmith** [comment]
First, don’t forget, you can and should learn how to do the thing Alex taught you how to do via git where you only upload the changed/brand new files.

**Comment by lepidopterane-atsmith** [comment]
_readme.md_
- It’s good to see you document updates in your readme!
- But the error that was always in your readmes is still there
- Plus, “Clarified error messages” is a lot less vague than “Modify some error messages”

**Comment by lepidopterane-atsmith** [comment]
_Chat.java_
- Your error message looks good here!
- The definition you use in lines 113 and 120 sounds vague and unsure, so you should just say “Get session information.”

**Comment by lepidopterane-atsmith** [comment]
_Context.java_
I understand what the code was trying to accomplish here, but it does not need an if-else to govern it. View.getInfo should always be called because it will return what needs to be returned, so if it returns a null it has printed the relevant error messages to explain that instance.

**Comment by lepidopterane-atsmith** [comment]
_View.java_
- Line 153 is nice and functional.
- I’m not sure about line 157, though. **Can you explain what was the reasoning behind that line? If it helps, maybe you should create a test file for tripping that error (If you do, include it in your next pull!)**

**Comment by lepidopterane-atsmith** [comment]
_ServerInfo.java_
Last pull, I explained why a try/catch statement was unnecessary in a ServerInfo constructor. This also goes for try/catch/finally.

File: README.md
Comment (aslo) [inline README.md]: I don't think this is the appropriate place to add this information. This README is focused on how to build and run the application, and should be kept as concise as possible. Information about new features being added belong in a Pull Request message, or commit message instead. 

I recommend deleting this `UPDATE FROM YUHANG LIAO` section of the README.

File: README.md
Comment (aslo) [inline README.md]: This might not have been intentional on your part, but you should change this line back to how it was originally.

File: README.md
Comment (aslo) [inline README.md]: Same as above, this might not have been intentional on your part, but you should change this line back to how it was originally.

File: src/codeu/chat/client/commandline/Chat.java
Comment (aslo) [inline src/codeu/chat/client/commandline/Chat.java]: Generally it's not a good idea to reference file names or line numbers in an error message like this, because in the future, someone might move the code around or rename the file, and forget to update this line. A better way to communicate this information to the user is to use an exception and a stack trace, which will always be up to date with the files and line numbers. 

If you want to make sure this information is included with the error message, try replacing this line with something like:
```
new IllegalStateException("Error: ServerInfo cannot be read.").printStackTrace();
```

File: src/codeu/chat/client/core/View.java
Comment (aslo) [inline src/codeu/chat/client/core/View.java]: Same issue here, remove file names and line numbers.

File: src/codeu/chat/client/core/View.java
Comment (aslo) [inline src/codeu/chat/client/core/View.java]: Same issue here, remove file names and line numbers.

File: src/codeu/chat/client/core/Context.java
Comment (aslo) [inline src/codeu/chat/client/core/Context.java]: Same issue as above - you shouldn't reference file names or line numbers like this in error messages. Consider changing this line. However, if you follow my comment to remove the class casting in this method, this line can just be deleted anyway.

File: src/codeu/chat/client/core/Context.java
Comment (aslo) [inline src/codeu/chat/client/core/Context.java]: Generally, you should avoid class casting like this if possible, since it has the opportunity to introduce runtime exceptions into the code. You should just change the declaration of this variable to look like `private final View view;` on line 29. Then this entire method can just be one line: 
```
return view.getInfo();
```

File: README.md
Comment (aslo) [reply README.md]: One other big issue - you've changed the file permissions of many files in this pull request. Notice that at the top, many files display a change like this:

screen shot 2017-05-29 at 2 56 22 pm

That means these files have been changed to an "executable" status, which is incorrect - they are not executable files, and this can be confusing to other developers in the future. More importantly, a pull request should only change the files relevant to the feature being worked on. Before merging this, I strongly recommend that you change these files back to the default permissions. Assuming you're on Linux or Mac, for each of these files, you should run `chmod 0644 <file_name_here>` to revert them back to the correct status.

[This stackoverflow answer](https://stackoverflow.com/a/8347325/5053647) has more info about file permissions in case you're interested.

**Review by aslo — CHANGES_REQUESTED** [review]
Nice job! I left you some comments and suggestions to fix before merging this pull request. Let me know if you have any questions!

File: src/codeu/chat/client/commandline/Chat.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/commandline/Chat.java]: printStackTrace doesn't take string parameters. Also what would be a stack trace for a new IOException?

File: src/codeu/chat/client/core/View.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/core/View.java]: 1. Why is this a fatal error?
2. An unexpected error could happen that would have nothing to do with the connection, but the one that got us here did. This line does not express that.
3. This line has the same error w/printStackTrace syntax as the previous line I corrected. It should be:
System.out.println("insert error message here");
ex.printStackTrace();

File: src/codeu/chat/common/ServerInfo.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/common/ServerInfo.java]: 1. What makes this error fatal?
2. Same syntax problem w/printStackTrace, different lines.
3. Second error sentence is unnecessary because nothing guaranteeable happens in the program after this message and its syntax is printed.

**Review by lepidopterane-atsmith — CHANGES_REQUESTED** [review]
You're getting better at this! I have a few questions, but really you just need to get a feel for how printStackTrace() works.

File: src/codeu/chat/common/ServerInfo.java
Comment (ComarPers922) [inline src/codeu/chat/common/ServerInfo.java]: Please notice I just use one line to print the error, which is printStackTrace.

Comment (ComarPers922) [reply src/codeu/chat/common/ServerInfo.java]: I will fix the parameter problem later.

File: src/codeu/chat/client/core/View.java
Comment (ComarPers922) [inline src/codeu/chat/client/core/View.java]: Please notice I just use one line to print the error, which is printStackTrace.

Comment (ComarPers922) [reply src/codeu/chat/client/core/View.java]: I really don't know how to explain an unexpected error...

File: README.md
Comment (lepidopterane-atsmith) [reply README.md]: ^^ What Alex said

File: README.md
Comment (lepidopterane-atsmith) [reply README.md]: ^^ What Alex said

File: src/codeu/chat/client/commandline/Chat.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/commandline/Chat.java:25]: This is just me being nitpicky but you should delete this line since it is a copy of line 17

File: src/codeu/chat/client/commandline/Chat.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/commandline/Chat.java]: "some information" is vague, and don't say anything after the semicolon. One way to do this is to replace that text with "Get session information."

File: src/codeu/chat/client/commandline/Chat.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/commandline/Chat.java:137]: "some information" is vague, and don't say anything after the semicolon. One way to do this is to replace that text with "Get the follwing session information: Version Info." Don't worry, that list will grow when Vicente has to merge his stuff with yours.

File: src/codeu/chat/client/commandline/Chat.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/commandline/Chat.java:146]: That format only works with try/catch(Exception e) scenarios; a simple System.out.println("ERROR: ServerInfo cannot be read."); would do. Also, when I ran this, I got a different error because there are no IOException constructors that just take a String.

File: src/codeu/chat/client/core/View.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/core/View.java:18]: We don't need this and I'll tell you why

File: src/codeu/chat/client/core/View.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/core/View.java:155]: That format only works with try/catch(Exception e) scenarios, and can only be implemented in the catch brackets. A simple System.out.println("ERROR: The server did not respond with the type of response we expected."); would do. Also, when I ran this, I got a different error because there are no IOException constructors that just take a String.

File: src/codeu/chat/client/core/View.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/core/View.java:159]: I understand what sort of format you are trying to do here, but here's the format in which it should be done:
catch (Exception ex) {
System.out.println("<insert error message here>");
ex.getStackTrace();

Also, it's important to note that this is not a fatal or unexpected error, or else the comments would have told us so.

File: src/codeu/chat/common/ServerInfo.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/common/ServerInfo.java:13]: SERVER_VERSION is final and static from the time it's initialized in the class, so it's a constant value that will never change. I should also point out that any time a constant like this is initialized, they usually have NAMES_LIKE_THIS . Because of these two things, a try/catch/finally is unnecessary in the constructor because if a value never changes to something it shouldn't be and is always initialized when it needs to be, it will be in a constant, non-erroneous state 24/7.

File: src/codeu/chat/client/commandline/Panel.java
Comment (lepidopterane-atsmith) [inline src/codeu/chat/client/commandline/Panel.java]: Good job with averting merge conflicts!

**Review by lepidopterane-atsmith — CHANGES_REQUESTED** [review]
Okay, because GitHub is special I reviewed your last two pulls (merging and previous changes because they are both significant). You did well with merging, but there are still some areas you need to work on. **Also, you should tell me about the last time you compiled and ran your commit.**

File: src/codeu/chat/common/ServerInfo.java
Comment (aslo) [reply src/codeu/chat/common/ServerInfo.java:13]: Just wanted to chime in here. A try/catch block is actually necessary here because the java compiler requires it. IOExceptions are a class of exceptions that we need to catch in our code, otherwise it just won't compile at all :)

**Review by aslo — CHANGES_REQUESTED** [review]
Hi @ComarPers922, this is looking really great! Thanks for addressing the feedback from my last review. I think this is almost good to merge! 

It looks like there are a few files whose permissions you still need to change back: 
* .travis.yml
* .gitignore
* README.md

Once you fix that, I'll be happy to approve this!

**Comment by lepidopterane-atsmith** [comment]
Unless Alex or anyone else has something to say here, LGTM.

**Pull request merged**
